### PR TITLE
ADAP-544: Add support for dynamic tables

### LINKED ISSUE #603: [ADAP-544] [Feature] Support Dynamic Tables as a materialization (following base Materialized View ergonomics)
For more context about Dynamic Tables (DTs), see [this post](https://www.snowflake.com/blog/dynamic-tables-delivering-declarative-streaming-data-pipelines/).

## `materialized: dynamic_table`

This issue is the local implementation of the core feature: [Let's add Materialized View as a materialization, finally](https://github.com/dbt-labs/dbt-core/issues/6911). 
The original spec is the baseline. In every aspect it's the same feature, with the exceptions described below.

### Alias

The local alias for this materialization is `dynamic_table`, not `materialized_view`. The later is not supported and should return an error (to avoid confusion with Snowflake's MVs)

### Schema management

Metadata caching/querying should be updated to add the following sources in order to support Dynamic Tables:
- table level via `SHOW DYNAMIC TABLES` - [doc](https://docs.snowflake.com/en/LIMITEDACCESS/show-dynamic-tables)
- column level via `DESCRIBE DYNAMIC TABLE` - [doc](https://docs.snowflake.com/LIMITEDACCESS/desc-dynamic-table)

### Configurations

Available parameters:
  - ~~LAG~~ TARGET_LAG : string (`<num> { seconds | minutes | hours | days }`)
  - WAREHOUSE : string
  - REFRESH_STRATEGY : {`no_wait`,`refreshed`}
  - From models: query_tag

Refresh strategy is described below.

**Out of scope** for the moment (not clear yet if applicable): [copy_grants](https://docs.getdbt.com/reference/resource-configs/snowflake-configs#copying-grants), [persist_docs](https://docs.getdbt.com/reference/resource-configs/persist_docs), [clustering](https://docs.getdbt.com/reference/resource-configs/snowflake-configs#configuring-table-clustering) management

### Lifecycles

- Lifecycle (`dbt run`) : 
  - **If** necessary backup (exists with different materialization)
  - **If** necessary `CREATE OR REPLACE DYNAMIC TABLE` (doesn't exist) and apply *refresh_strategy* **Else** 
    - Check status (see below)
      - **If** `FAILED` then `CREATE OR REPLACE DYNAMIC TABLE` and apply *refresh_strategy* **Else** 
        - **If** the options of the object in the database are different from the current model, depending on `on_configuration_change`
        - `skip` : Do nothing, aka stage changes until the next full refresh
        - `fail` : Fail the run
        - `apply` : [default] Apply the changes
 - Lifecycle (`dbt run --full-refresh`) :
   - **If** necessary backup (exists with different materialization ~~OR exists and full-refresh~~)
   - `CREATE OR REPLACE DYNAMIC TABLE` and apply *refresh_strategy*

### Status check

The refresh status of a DT is obtained via the following query:

```SQL 
SELECT TOP 1 state
FROM TABLE(INFORMATION_SCHEMA.DYNAMIC_TABLE_REFRESH_HISTORY(NAME => '...'))
ORDER BY refresh_version DESC;
```

The return code is a string [among](https://docs.snowflake.com/LIMITEDACCESS/dynamic_table_refresh_history#output):

> SCHEDULED, EXECUTING, SUCCEEDED, FAILED, CANCELLED, or SKIPPED

### Refresh strategy

To ensure downstream models are able to query the DT in a proper state, dbt may have to wait for the DT to be successfully refreshed before moving on. The `refresh_strategy` setting offers 2 options:

|Refresh strategy|Behavior|
|-|-|
|`no_wait`|default: dbt will not wait, it will proceed as soon as the DDL statement is issued|
|`refresh`|optional: dbt will wait until the DT is in the `SUCCEEDED` state, polling every few seconds its status|

In order to avoid waiting indefinitely, in `refresh` mode, the polling will stop and the **run will fail** when:
- after the TARGET_LAG period has lapsed, any state other than `SCHEDULED, EXECUTING` is returned
- after 2 TARGET_LAG periods have lapsed

### PR DESCRIPTION
**Pull request by mikealfare**
resolves #603

### Description

- Add support for dynamic tables
- Follow the pattern for dbt materialized views (not snowflake materialized views)
- Ensure exception is thrown for "materialized_view" materialization to avoid confusion

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-snowflake/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-snowflake/blob/main/CONTRIBUTING.md#Adding-CHANGELOG-Entry)

### CODE REVIEW

File: dbt/adapters/snowflake/relation_configs/dynamic_table.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+from dbt.adapters.relation_configs import RelationConfigBase
+
+from dbt.adapters.snowflake.relation_configs.lag import SnowflakeDynamicTableLagConfig
+
+
+@dataclass(frozen=True, eq=True, unsafe_hash=True)
+class SnowflakeDynamicTableConfig(RelationConfigBase):
+    """
+    This config follow the specs found here:
+    https://docs.snowflake.com/en/LIMITEDACCESS/create-dynamic-table
```
Comment (mikealfare) [inline dbt/adapters/snowflake/relation_configs/dynamic_table.py:12]: Remove the URL

File: dbt/include/snowflake/macros/materializations/dynamic_table/ddl.sql [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+{% macro snowflake__get_replace_dynamic_table_as_sql(relation, sql, existing_relation, backup_relation, intermediate_relation) -%}
+    {{- log('Applying REPLACE to: ' ~ relation) -}}
+    {{ snowflake__get_drop_dynamic_table_sql(existing_relation) }};
+    {{ snowflake__get_create_dynamic_table_as_sql(relation, sql) }}
+{%- endmacro %}
+
+
+{% macro snowflake__refresh_dynamic_table(relation) -%}
+    {{- log('Applying REFRESH to: ' ~ relation) -}}
+    alter dynamic table {{ relation }} set lag = '{{ config.get("lag") }}'
```
Comment (mikealfare) [inline dbt/include/snowflake/macros/materializations/dynamic_table/ddl.sql:34]: Replace this with a no-op

File: dbt/include/snowflake/macros/materializations/dynamic_table/materialization.sql [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        -- get config options
+        {% set on_configuration_change = config.get('on_configuration_change') %}
+        {% set configuration_changes = snowflake__get_dynamic_table_configuration_changes(existing_relation, config) %}
+
+        {% if configuration_changes is none %}
+            {% set build_sql = snowflake__refresh_dynamic_table(target_relation) %}
+
+        {% elif on_configuration_change == 'apply' %}
+            {% set build_sql = snowflake__get_alter_dynamic_table_as_sql(target_relation, configuration_changes, sql, existing_relation, backup_relation, intermediate_relation) %}
+        {% elif on_configuration_change == 'skip' %}
```
Comment (mikealfare) [inline dbt/include/snowflake/macros/materializations/dynamic_table/materialization.sql:77]: Continue

File: dbt/include/snowflake/macros/materializations/dynamic_table/materialization.sql [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+
+{% macro dynamic_table_get_build_sql(existing_relation, target_relation, backup_relation, intermediate_relation) %}
+
+    {% set full_refresh_mode = should_full_refresh() %}
+
+    -- determine the scenario we're in: create, full_refresh, alter, refresh data
+    {% if existing_relation is none %}
+        {% set build_sql = snowflake__get_create_dynamic_table_as_sql(target_relation, sql) %}
+    {% elif full_refresh_mode or not existing_relation.is_view %}
```
Comment (mikealfare) [inline dbt/include/snowflake/macros/materializations/dynamic_table/materialization.sql:64]: is_materialized_view

File: dbt/include/snowflake/macros/materializations/dynamic_table/materialization.sql [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+            {% set build_sql = snowflake__get_alter_dynamic_table_as_sql(target_relation, configuration_changes, sql, existing_relation, backup_relation, intermediate_relation) %}
+        {% elif on_configuration_change == 'skip' %}
+            {% set build_sql = '' %}
+            {{ exceptions.warn("Configuration changes were identified and `on_configuration_change` was set to `skip` for `" ~ target_relation ~ "`") }}
+        {% elif on_configuration_change == 'fail' %}
+            {{ exceptions.raise_compiler_error("Configuration changes were identified and `on_configuration_change` was set to `fail` for `" ~ target_relation ~ "`") }}
+
+        {% else %}
+            -- this only happens if the user provides a value other than `apply`, 'skip', 'fail'
+            {{ exceptions.raise_compiler_error("Unexpected configuration scenario") }}
```
Comment (colin-rogers-dbt) [inline dbt/include/snowflake/macros/materializations/dynamic_table/materialization.sql:85]: nit: Include `on_configuration_change` value in exception message?

File: dbt/include/snowflake/macros/materializations/dynamic_table/materialization.sql [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        {% elif on_configuration_change == 'apply' %}
+            {% set build_sql = snowflake__get_alter_dynamic_table_as_sql(target_relation, configuration_changes, sql, existing_relation, backup_relation, intermediate_relation) %}
+        {% elif on_configuration_change == 'skip' %}
+            {% set build_sql = '' %}
+            {{ exceptions.warn("Configuration changes were identified and `on_configuration_change` was set to `skip` for `" ~ target_relation ~ "`") }}
+        {% elif on_configuration_change == 'fail' %}
+            {{ exceptions.raise_compiler_error("Configuration changes were identified and `on_configuration_change` was set to `fail` for `" ~ target_relation ~ "`") }}
+
+        {% else %}
+            -- this only happens if the user provides a value other than `apply`, 'skip', 'fail'
```
Comment (mikealfare) [inline dbt/include/snowflake/macros/materializations/dynamic_table/materialization.sql:84]: continue

File: tests/functional/adapter/dynamic_table_tests/test_dynamic_table.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+        # poll database
+        table_end = get_row_count(project, "base_table")
+        dyn_end = get_row_count(project, "base_dynamic_table")
+
+        # new records were inserted in the table but didn't show up in the dynamic table until it was refreshed
+        assert table_start < table_mid == table_end
+        assert dyn_start == dyn_mid < dyn_end
+
+        # record counts tie at start an end, and only differ in the middle due to refresh timing
+        assert table_start == dyn_start
```
Comment (mikealfare) [inline tests/functional/adapter/dynamic_table_tests/test_dynamic_table.py:73]: Move this up

File: tests/functional/adapter/dynamic_table_tests/test_dynamic_table.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+        assert table_end == dyn_end
+
+
+@pytest.mark.skip("We're not looking for changes yet")
```
Comment (mikealfare) [inline tests/functional/adapter/dynamic_table_tests/test_dynamic_table.py:77]: Add TODO that this is under active development (why are these here?)

**Review by colin-rogers-dbt — APPROVED** [review]

**Comment by alexandrewatercare** [comment]
Hello, how are we supposed to use the the dynamic0 table configuration, I try to use it but it seems not recognize the warehouse tag.
models:
  - name: dw_ips_dim_date_calendar
    description: "Dimensional date table"
    config:
      materialized: dynamic_table
      target_lag: "60 minutes"
      warehouse: "wh_dynamic_table"

Snowflake adapter: Snowflake error: 090080 (22000): Warehouse 'NONE' does not exist.
